firebase-installations/src/main/java/com/google/firebase/installations/local/IidStore.java

```diff
@@ -0,0 +1,110 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.installations.local;
+
+import static android.content.ContentValues.TAG;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.util.Base64;
+import android.util.Log;
+import androidx.annotation.GuardedBy;
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import com.google.firebase.FirebaseApp;
+import java.security.KeyFactory;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.PublicKey;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.X509EncodedKeySpec;
+
+public class IidStore {
+  private static final String IID_SHARED_PREFS_NAME = "com.google.android.gms.appid";
+  private static final String STORE_KEY_PUB = "|S||P|";
+  private static final String STORE_KEY_ID = "|S|id";
+
+  @GuardedBy("iidPrefs")
+  private final SharedPreferences iidPrefs;
+
+  public IidStore(@NonNull FirebaseApp firebaseApp) {
+    // Different FirebaseApp in the same Android application should have the same application
+    // context and same dir path. We only read existing Iids for the default firebase application.
+    iidPrefs =
+        firebaseApp
+            .getApplicationContext()
+            .getSharedPreferences(IID_SHARED_PREFS_NAME, Context.MODE_PRIVATE);
+  }
+
+  @Nullable
+  public String readIid() {
+    synchronized (iidPrefs) {
+      // Reading iid from shared preferences
+      String id = iidPrefs.getString(STORE_KEY_ID, /* defaultValue= */ null);
+
+      if (id != null) {
+        return id;
+      }
```

Comment:
(Feel free to ignore this)
How about outsourcing this into its own method, even though its only one line:

private Optional\<String> readInstanceIdFromLocalStorage() {
  return Optional.ofNullable(iidPrefs.getString(STORE_KEY_ID, /* defaultValue= */ null));
}

(after reading my comments below)
You can then rewrite this method like this:

return readInstanceIdFromLocalStorage().orElse(this:: readPublicKeyFromLocalStorageAndCalculateInstanceIdFromIt);